clap_complete/src/dynamic/completer.rs

```diff
@@ -708,3 +731,75 @@ impl CompletionCandidate {
         self.visible
     }
 }
+
+/// This trait is used to provide users a way to add custom value hint to the argument.
+/// This is useful when predefined value hints are not enough.
+pub trait CustomCompleter: core::fmt::Debug + Send + Sync {
+    /// This method should return a list of custom value completions.
+    /// If there is no completion, it should return `vec![]`.
+    ///
+    /// See [`CompletionCandidate`] for more information.
+    fn custom_hint(&self) -> Vec<CompletionCandidate>;
+}
+
+/// A wrapper for custom completer
+///
+/// # Example
+///
+/// ```rust
+/// use clap_complete::dynamic::{ArgValueCompleter, CustomCompleter};
+/// use clap_complete::dynamic::CompletionCandidate;
+/// use std::ffi::OsString;
+///
+/// #[derive(Debug)]
+/// struct MyCustomCompleter {
+///     file: std::path::PathBuf,
+/// }
+///
+/// impl CustomCompleter for MyCustomCompleter {
+///     fn custom_hint(&self) -> Vec<CompletionCandidate> {
+///         let content = std::fs::read_to_string(&self.file);
+///         match content {
+///             Ok(content) => {
+///                 content.lines().map(|os| CompletionCandidate::new(os).visible(true)).collect()
+///             }
+///             Err(_) => vec![],
+///         }
+///     }
+/// }
+///
+/// fn main() {
+///     let completer = ArgValueCompleter::new(MyCustomCompleter{
+///         file: std::path::PathBuf::from("/path/to/file"),
+///     });
+///
+///     // TODO: Need to implement this command by using derive API.
+///     // This is just a placeholder to show how to add custom completer.
+///     let mut arg = clap::Arg::new("custom").long("custom");
+///     arg.add(completer);
+///     let mut cmd = clap::Command::new("dynamic").arg(arg);
+/// }
+///    
+/// ```
+#[derive(Debug, Clone)]
+pub struct ArgValueCompleter(Arc<dyn CustomCompleter>);
+
+impl ArgValueCompleter {
+    /// Create a new `ArgValueCompleter` with a custom completer
+    pub fn new<C: CustomCompleter>(completer: C) -> Self
+    where
+        C: 'static + CustomCompleter,
+    {
+        Self(Arc::new(completer))
+    }
+}
+
+impl Deref for ArgValueCompleter {
+    type Target = dyn CustomCompleter;
+
+    fn deref(&self) -> &Self::Target {
+        &*self.0
+    }
+}
```

Comment:
Instead should we have `ArgValueCompleter` implement `ArgValueCompleter`?

Reply:
Thanks for your review! This is just my personal habit to make it more convenient when using internal fields, but it also exposes the inner of `ArgValueCompleter` to users :worried:.